### PR TITLE
[Case 6491] Remainder of entitySelectionTool clean up (details below).

### PR DESCRIPTION
**Pull request by 1P-Cusack**
This pull request contains the remainder, Parts 2 - 6, of the entitySelectionTool cleanup for [FogBugz Case 6491](https://highfidelity.fogbugz.com/f/cases/6491/edit-js-work-proposed-by-1st-playable).  Part 1 was merged via PR #11165.  This pull request also replaces https://github.com/highfidelity/hifi/pull/11205, as mentioned in https://github.com/highfidelity/hifi/pull/11205#issuecomment-328642626

- Fixes the issue where grabbers would re-appear when rotating the selected object rather than staying hidden.
- Fixes a bug with translateXZTool where its startingElevation and startingDistance properties were only set when local debug var was set. This appears to have been responsible for being able to move the object farther than was visible. Re-tested with fix and wasn't able to get that behavior any longer.
- This fixes an issue with multi-selection object rotation for pitch & roll that was previously exclusive to yaw rotation tool and also dedupes rotation tool code between the 3 tools where able.
- Fixes situations where attempting to click on a rotate, grab, or scale handle that was in front of others might unexpectedly activate the obscured handle.
- Fixes rotation skip near 0 and 180 degrees "poles"
- Fixes issue where the user view was obstructed by grabber handles when translating the selected object(s).
- Cleaned up onMouseEvents and moved all functionality for tools within the tools themselves as opposed to being split between the tool & which ever mouse event.

For more details, please refer to contained commit messages.

**Testing:**

1. Enter a domain where you have editing/rez permissions.
2. Create and/or select objects and perform the following operations with both Single & Multi-selection:
    - Rotation
    - Translation
    - Scaling
    - Cloning
3. The behavior should be inline with the expected results below.

**Expected Behavior:**

**Rotation:**
- While rotating longer the outer circle overlay, the user should see fine grain degree changes even around the 0 & 180 degree points.
- Multi-selection rotation along the Pitch & Roll, like Yaw, should now have proper solar system type rotation. (Note:  The "sun" position is the center of bounding box which encapsulates _all_ items within the selection).
- Upon clicking a rotation handle, all rotation, translation, scaling, & grabber handles should turn and remain invisible until the user releases the mouse press.

**Translation:**
- When moving the object along the z-axis:
    - The object doesn't warp behind the user/camera when moving the selection towards themselves or into infinity when moving the selection forward into the screen.
    - All rotation, translation, scaling, & grabber handles should turn invisible on click and remain invisible until the user releases the mouse press.
- When moving the object along the x-axis:
    - It's no longer possible to warp the selection into infinity.
    - All rotation, translation, scaling, & grabber handles should turn invisible on click and remain invisible until the user releases the mouse press.
- When moving the object along the y-axis:
    - It's no longer possible to warp the selection into infinity.
    - The vertical translation handle should remain visible when clicked and stay visible during the translation.  When the user releases the mouse press, the vertical translation handle should remain visible.
    - Excluding the vertical translation handle, all other rotation, translation, scaling, & grabber handles should turn invisible on clicking the vertical translation handle.  All the invisible handles should remain invisible until the user releases the mouse press.

**Multi-Selection:**
- All translation and rotation behavior noted above should also apply to multi-selection.

**General:**
- When clicking on a pair of overlapping handles, the handle which is directly under the cursor/ray should activate as opposed to the last tool checked.

### CODE REVIEW

**Comment by 1P-Cusack** [comment]
@themelissabrown & @misslivirose:  Here's the replacement pull request for https://github.com/highfidelity/hifi/pull/11205 as discussed.  This contains the remainder of the changes for [FogBugz Case 6491](https://highfidelity.fogbugz.com/f/cases/6491/edit-js-work-proposed-by-1st-playable).

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/3209/HighFidelity-Beta-PR11338-81540aa6baab7c1108a6eefa7c2f5cccb39c5f84.dmg

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/3209/HighFidelity-Beta-PR11338-81540aa6baab7c1108a6eefa7c2f5cccb39c5f84.exe

**Comment by misslivirose** [comment]
Thanks @1P-Cusack - I'll take a look today

File: scripts/system/libraries/entitySelectionTool.js [hunk 1]
```diff
@@ -1833,264 +1846,164 @@ SelectionDisplay = (function() {
...
+            print("    Set Non-Light Grabbers Visible - Norm: " + stretchHandlesVisible + " Ext: " + extendedStretchHandlesVisible);
+        }
+        var isSingleSelection = (selectionManager.selections.length == 1);
 
-        if (selectionManager.selections.length == 1) {
+        if (isSingleSelection) {
             var properties = Entities.getEntityProperties(selectionManager.selections[0]);
-            if (properties.type == "Light" && properties.isSpotlight) {
+            var isLightSelection = (properties.type == "Light");
+            if ( isLightSelection ) {
```
Comment (ZappoMan) [inline scripts/system/libraries/entitySelectionTool.js]: coding standard: this spacing is weird, it should be `if (isLightSelection) {`

File: scripts/system/libraries/entitySelectionTool.js [hunk 2]
```diff
@@ -1833,264 +1846,164 @@ SelectionDisplay = (function() {
...
+                if (wantDebug) {
+                    print("    Light Selection revoking Non-Light Grabbers Visibility!");
+                }
                 stretchHandlesVisible = false;
                 extendedStretchHandlesVisible = false;
+                cloneHandleVisible = false;
+                if (properties.isSpotlight) {
+                    that.setPointLightHandlesVisible( false );
```
Comment (ZappoMan) [inline scripts/system/libraries/entitySelectionTool.js]: coding standard: spacing wrong

File: scripts/system/libraries/entitySelectionTool.js [hunk 3]
```diff
@@ -1833,264 +1846,164 @@ SelectionDisplay = (function() {
...
-                    visible: false
-                });
-                Overlays.editOverlay(grabberSpotLightLineT, {
-                    visible: false
-                });
-                Overlays.editOverlay(grabberSpotLightLineB, {
-                    visible: false
-                });
+                } else { //..it's a PointLight
+                    that.setSpotLightHandlesVisible( false );
```
Comment (ZappoMan) [inline scripts/system/libraries/entitySelectionTool.js]: weird spacing

File: scripts/system/libraries/entitySelectionTool.js [hunk 4]
```diff
@@ -1833,264 +1846,164 @@ SelectionDisplay = (function() {
...
-                Overlays.editOverlay(grabberPointLightR, {
-                    visible: false
-                });
-                Overlays.editOverlay(grabberPointLightF, {
-                    visible: false
-                });
-                Overlays.editOverlay(grabberPointLightN, {
-                    visible: false
-                });
+                    var showEdgePointGrabbers = ! inModeTranslate;
```
Comment (ZappoMan) [inline scripts/system/libraries/entitySelectionTool.js]: no space after `!`

File: scripts/system/libraries/entitySelectionTool.js [hunk 5]
```diff
@@ -2312,11 +2225,11 @@ SelectionDisplay = (function() {
             z: position.z
         };
         Overlays.editOverlay(grabberMoveUp, {
-            visible: (activeTool === null) || (mode == "TRANSLATE_UP_DOWN")
+            visible: ( ! activeTool ) || isActiveTool(grabberMoveUp)
```
Comment (ZappoMan) [inline scripts/system/libraries/entitySelectionTool.js]: fix spacing `!activeTool || isActiveTool(grabberMoveUp)`

File: scripts/system/libraries/entitySelectionTool.js [hunk 6]
```diff
@@ -2330,26 +2243,69 @@ SelectionDisplay = (function() {
...
-                visible: isVisible
-            });
-        }
-        length = selectionBoxes.length;
-        for (var boxIndex = 0; boxIndex < length; boxIndex++) {
-            Overlays.editOverlay(selectionBoxes[boxIndex], {
-                visible: isVisible
-            });
-        }
+        helperSetOverlaysVisibility( allOverlays, isVisible );
```
Comment (ZappoMan) [inline scripts/system/libraries/entitySelectionTool.js]: spacing

File: scripts/system/libraries/entitySelectionTool.js [hunk 7]
```diff
@@ -2330,26 +2243,69 @@ SelectionDisplay = (function() {
...
+    that.setRotationHandlesVisible = function(isVisible) {
+        var visibilityUpdate = { visible: isVisible };
+        Overlays.editOverlay(yawHandle, visibilityUpdate);
+        Overlays.editOverlay(pitchHandle, visibilityUpdate);
+        Overlays.editOverlay(rollHandle, visibilityUpdate);
+    };
+
+    // FUNCTION: SET STRETCH HANDLES VISIBLE
+    that.setStretchHandlesVisible = function(isVisible) {
+        helperSetOverlaysVisibility( stretchHandles, isVisible );
```
Comment (ZappoMan) [inline scripts/system/libraries/entitySelectionTool.js]: spacing

File: scripts/system/libraries/entitySelectionTool.js [hunk 8]
```diff
@@ -2330,26 +2243,69 @@ SelectionDisplay = (function() {
...
+
+    // FUNCTION: SET GRABBER MOVE UP VISIBLE
+    that.setGrabberMoveUpVisible = function(isVisible) {
+        Overlays.editOverlay(grabberMoveUp, { visible: isVisible });
+    };
+
+    // FUNCTION: SET GRABBER TOOLS UP VISIBLE
+    that.setGrabberToolsVisible = function(isVisible) {
+        var visibilityUpdate = { visible: isVisible };
+        for ( var toolKey in grabberTools ) {
```
Comment (ZappoMan) [inline scripts/system/libraries/entitySelectionTool.js]: spacing

File: scripts/system/libraries/entitySelectionTool.js [hunk 9]
```diff
@@ -2330,26 +2243,69 @@ SelectionDisplay = (function() {
...
+            if ( ! grabberTools.hasOwnProperty( toolKey ) ) {
```
Comment (ZappoMan) [inline scripts/system/libraries/entitySelectionTool.js]: spacing

File: scripts/system/libraries/entitySelectionTool.js [hunk 10]
```diff
@@ -2330,26 +2243,69 @@ SelectionDisplay = (function() {
...
+    };
+
+    // FUNCTION: SET POINT LIGHT HANDLES VISIBLE
+    that.setPointLightHandlesVisible = function(isVisible) {
+        helperSetOverlaysVisibility( pointLightGrabberHandles, isVisible );
+    };
+
+    // FUNCTION: SET SPOT LIGHT HANDLES VISIBLE
+    that.setSpotLightHandlesVisible = function(isVisible) {
+        helperSetOverlaysVisibility( spotLightGrabberHandles, isVisible );
```
Comment (ZappoMan) [inline scripts/system/libraries/entitySelectionTool.js]: spacing

**Comment by 1P-Cusack** [comment]
@ZappoMan: Spacing issues have been addressed.

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/3284/HighFidelity-Beta-PR11338-e06f46876a9b5f21594206502d0843340c6cce78.exe

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/3284/HighFidelity-Beta-PR11338-e06f46876a9b5f21594206502d0843340c6cce78.dmg

**Comment by Nex-Pro** [comment]
Testing this

**Comment by Nex-Pro** [comment]
Testing:

Enter a domain where you have editing/rez permissions.
Create and/or select objects and perform the following operations with both Single & Multi-selection:
Rotation
Translation
Scaling
Cloning
The behavior should be inline with the expected results below.
Expected Behavior:

### Rotation: **FAIL** In HMD If the rotation handle is low and between the ent and the avatar the rotation graticle shows and is active ( red ) but no rotation is possible.

While rotating longer the outer circle overlay, ( will assume longer is along )
the user should see fine grain degree changes even around the 0 & 180 degree points. **SUCCESS**

Multi-selection rotation along the Pitch & Roll, 
like Yaw, should now have proper solar system type rotation.   **SUCCESS**
(Note: The "sun" position is the center of bounding box which encapsulates all items within the selection).

Upon clicking a rotation handle, all rotation, translation, scaling, 
& grabber handles should turn and remain invisible until the user releases the mouse press.  **SUCCESS**


### Translation:

When moving the object along the z-axis:
The object doesn't warp behind the user/camera when moving the selection towards themselves 
or into infinity when moving the selection forward into the screen.  **SUCCESS**

All rotation, translation, scaling, 
& grabber handles should turn invisible on click and remain invisible until the user releases the mouse press.
When moving the object along the x-axis:  **SUCCESS**

It's no longer possible to warp the selection into infinity.  **SUCCESS**
All rotation, translation, scaling, & grabber handles should turn invisible on click and remain invisible until the user releases the mouse press.


When moving the object along the y-axis:
It's no longer possible to warp the selection into infinity. **SUCCESS**

The vertical translation handle should remain visible when clicked and stay visible during the translation. **SUCCESS**
When the user releases the mouse press, the vertical translation handle should remain visible. **SUCCESS**

Excluding the vertical translation handle, 
all other rotation, translation, scaling, & grabber handles should turn invisible on clicking the vertical translation handle. **SUCCESS**
All the invisible handles should remain invisible until the user releases the mouse press. **SUCCESS**


### Multi-Selection:

All translation and rotation behavior noted above should also apply to multi-selection. **SUCCESS**


### General:

When clicking on a pair of overlapping handles, 
the handle which is directly under the cursor/ray should activate as opposed to the last tool checked. **SUCCESS** See notes re this.

### Notes:

There's a great deal of improvement here. In general this is good.

There seems to be a rendering issue with the control handles that deprives the desktop user of a sense of depth order. Look at the two overlapping control handles in the center of the purple cube. The lighting related brightness of the upper surface is over written by the cube behind. As a result the two handles appear to be one irregular solid. See image:

The way that translation in the avatar's Z axis is limited can be really annoying. IMO this needs a rethink. Try this with a cube of 500, 500, 500 First try it low and in front of you. It will stop after a seemingly short distance.  Then move it up some amount then move in the avatar Z direction and see the cube vanish due to lod distance.


Critical design flaw:

From a design point of view I can't see any good reason to scale the rotation graticle relative to the entities. This makes it unuseable for tiny or huge entities.

To repro:

Make a tiny cube or sphere less than 1cm.  Then try to rotate it or in fact to do anything with it.

Make a huge cube. enjoy.

**Comment by 1P-Cusack** [comment]
@Nex-Pro:

🙂 Yes, "While rotating longer...." should've been "While rotating along..."

Is it possible to get a screen capture or video of rotation scenario you encountered within the HMD where the ui shows; however, you can't rotate?  It sounds like something that can be broken out into its own issue as opposed to blocking this pull request.

Scaling the rotation ui relative to the entities is pre-existing behavior, and wasn't changed or influenced within this pull request.  Similar behavior can be seen with the rotation, scaling, grabbing, translation handles as well.  It was something that was noticed during this refactor and noted along with some other potential things to take a second look at to improve usability.  I tried looking through the history to see if there was any information regarding the choice of scaling that ui; however, I don't recall finding anything.  It may have been a byproduct of consistency with what was in place for the scaling handles, if they didn't go in around the same time.

This pass of the script was to address some of the more blatant issues encountered like the degree snapping, at times seeming unresponsiveness of certain actions, and also do clean up pass on the script to improve the ease of maintenance with regard to future updates or improvements.

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/3742/HighFidelity-Beta-PR11338-ff051db79a2ed32d8629f9b087f9f39fc6223d28.dmg

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/3742/HighFidelity-Beta-PR11338-ff051db79a2ed32d8629f9b087f9f39fc6223d28.exe

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/3770/HighFidelity-Beta-PR11338-6571aef997acdb466a9bafcd2528806be357d437.dmg

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/3770/HighFidelity-Beta-PR11338-6571aef997acdb466a9bafcd2528806be357d437.exe

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/3802/HighFidelity-Beta-PR11338-cb264b5552594ebb552972b14a45fb042dc60c68.dmg

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/3802/HighFidelity-Beta-PR11338-cb264b5552594ebb552972b14a45fb042dc60c68.exe

File: scripts/system/libraries/entitySelectionTool.js [hunk 11]
```diff
@@ -235,8 +238,14 @@ SelectionManager = (function() {
 
 // Normalize degrees to be in the range (-180, 180]
 function normalizeDegrees(degrees) {
-    while (degrees > 180) degrees -= 360;
-    while (degrees <= -180) degrees += 360;
+    while (degrees > 180) {
+     degrees -= 360;
```
Comment (ZappoMan) [inline scripts/system/libraries/entitySelectionTool.js]: this does not match coding standard of 4 spaces per indent

File: scripts/system/libraries/entitySelectionTool.js [hunk 12]
```diff
@@ -1529,34 +1478,31 @@ SelectionDisplay = (function() {
...
-        if (mode == "ROTATE_YAW" || mode == "ROTATE_PITCH" || mode == "ROTATE_ROLL" || mode == "TRANSLATE_X in case they Z") {
+        if (isActiveTool(yawHandle) || isActiveTool(pitchHandle) || 
+                isActiveTool(rollHandle) || isActiveTool(selectionBox) || isActiveTool(grabberCloner)) {
             rotationOverlaysVisible = true;
             rotateHandlesVisible = false;
-            translateHandlesVisible = false;
-            stretchHandlesVisible = false;
-            selectionBoxVisible = false;
-        } else if (mode == "TRANSLATE_UP_DOWN" || isPointLight) {
+            // translateHandlesVisible = false;
```
Comment (ZappoMan) [inline scripts/system/libraries/entitySelectionTool.js]: coding standard says not to leave dead code... why is this commented out?

Comment (1P-Cusack) [reply scripts/system/libraries/entitySelectionTool.js]: There's a notation where the commented out vars were declared regarding why those are left commented out.  They remain as they documented as they document intent of state as it
relates to modes that may be useful later.

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/3896/HighFidelity-Beta-PR11338-a5b913a8b992f73216cc2c4af7c308deb85b9b09.dmg

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/3896/HighFidelity-Beta-PR11338-a5b913a8b992f73216cc2c4af7c308deb85b9b09.exe

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/3906/HighFidelity-Beta-PR11338-18d884f6c45c00cc4086fa2472139bd6d82d7afc.dmg

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/3906/HighFidelity-Beta-PR11338-18d884f6c45c00cc4086fa2472139bd6d82d7afc.exe

File: scripts/system/libraries/entitySelectionTool.js [hunk 13]
```diff
@@ -235,8 +238,14 @@ SelectionManager = (function() {
...
 function normalizeDegrees(degrees) {
-    while (degrees > 180) degrees -= 360;
-    while (degrees <= -180) degrees += 360;
+    while (degrees > 180) {
+        degrees -= 360;
+    }
+
+    while (degrees <= -180) {
+        degrees += 360;
+    }
```
Comment (AndrewMeadows) [inline scripts/system/libraries/entitySelectionTool.js]: The new code is cleaner and technically correct, but the method remains inefficient for large values of `degrees` and makes me cry.  Not a blocker but for bonus points: please replace this with:
```
function normalizeDegrees(degrees) {
    degrees = ((degrees + 180) % 360) - 180;
    if (degrees < -180) {
        degrees += 360;
    }
    return degrees;
}
```

Comment (1P-Cusack) [reply scripts/system/libraries/entitySelectionTool.js]: :nods: I was going with the approach of least amount of change to eliminate some trivial eslint issues present in this function as it was some original code that I'd not had reason to fiddle with; however, I'll upstream the change shortly.

**Comment by 1P-Cusack** [comment]
@AndrewMeadows:  I just committed the suggested change from https://github.com/highfidelity/hifi/pull/11338#pullrequestreview-67520585.  Once it builds, it should be good to go.

**Comment by 1P-Cusack** [comment]
@AndrewMeadows:  I just committed a slight modification to the suggested code to account for the range `(-180, 180]`.  `< -180` is now `<= 180` to trigger the adjustment behavior when degrees is `-180`.  Once that change builds, it should be good.

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/3941/HighFidelity-Beta-PR11338-2f082d9e86684a791e928750a21863084af47fdf.dmg

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/3941/HighFidelity-Beta-PR11338-2f082d9e86684a791e928750a21863084af47fdf.exe

**Comment by Nex-Pro** [comment]
T E S T I N G  Win10

**Comment by Nex-Pro** [comment]
## PASS with some design issues noted.

### Testing platform:
Intel i7 workstation.
gtx 1070 GPU
16gb ram.
Win 10.


Testing:

Enter a domain where you have editing/rez permissions.

Create and/or select objects and perform the following operations with both Single & Multi-selection:

Rotation
Translation
Scaling
Cloning

The behavior should be inline with the expected results below.


## Expected Behavior:

### Rotation:

While rotating longer the outer circle overlay, 
the user should see fine grain degree changes even around the 0 & 180 degree points. **SUCCESS** But read issues at bottom.

Multi-selection rotation along the Pitch & Roll, like Yaw, should now have proper solar system type rotation. 
(Note: The "sun" position is the center of bounding box which encapsulates all items within the selection). **SUCCESS**

Upon clicking a rotation handle, all rotation, translation, scaling, & grabber handles should turn and remain invisible until the user releases the mouse press. **SUCCESS**


### Translation:

#### When moving the object along the z-axis:

The object doesn't warp behind the user/camera when moving the selection towards themselves **There is an issue where the avatar is moved by the entities when the avatar is in third person. Otherwise this is OK.**

or into infinity when moving the selection forward into the screen. **SUCCESS**

All rotation, translation, scaling, & grabber handles should turn invisible on click and remain invisible until the user releases the mouse press. **SUCCESS**


#### When moving the object along the x-axis:

It's no longer possible to warp the selection into infinity. **SUCCESS**

All rotation, translation, scaling, & grabber handles should turn invisible on click and remain invisible until the user releases the mouse press. **SUCCESS**


#### When moving the object along the y-axis:

It's no longer possible to warp the selection into infinity. **SUCCESS** Movement is restricted to be within the screen-space

The vertical translation handle should remain visible when clicked and stay visible during the translation. When the user releases the mouse press, the vertical translation handle should remain visible. **SUCCESS**

Excluding the vertical translation handle, all other rotation, translation, scaling, & grabber handles should turn invisible on clicking the vertical translation handle. All the invisible handles should remain invisible until the user releases the mouse press. **SUCCESS**

### Multi-Selection:

All translation and rotation behavior noted above should also apply to multi-selection. **SUCCESS**


### General:

When clicking on a pair of overlapping handles, the handle which is directly under the cursor/ray should activate as opposed to the last tool checked. **SUCCESS** But read relevant note below.



### Issues Noted:

Rotation:

    The rotation reticles are occluded by entities including non-selected entities. 
    This has the effect of hiding the reticle in many situations.
    There is no readout of the angle that is being selected.

Handles:

    There are still situations where it is only possible to grab a particular handle by first changing the avatar's position/viewpoint.

Scaling:

    When scaling an entity using one of the corner handles there can be occasions where the scale suddenly jumps. Particularly where the entity occupies a large portion of the view.

### Green handle cannot be grabbed without changing avatar's viewpoint.
hifi-snap-by--on-2017-10-09_21-50-02

### Reticle is occluded.
hifi-snap-by--on-2017-10-09_21-10-47

### Rotation reticle is completely obscured
hifi-snap-by--on-2017-10-09_21-04-05

**Pull request merged**